resources: Add exceptions if the resource JSON has schema issues

Harshil2107 opened the pull request:
- This change adds exceptions and error checks to the resources JSON object. Exceptions are raised with a message if the resources have schema issues.

- This change also adds unit tests to test the new exceptions.

Harshil2107 committed:
resources: add exceptions for invalid resource schema

- This commit adds Exception if the resource dictionary has fields that are incorrect.

Harshil2107 committed:
tests: Add unit tests for new exceptions in resources

- This commit adds unit tests that tests the behaviour of gem5 if resoruces json have schema errors

Harshil2107 committed:
resources: remove extra print statement

powerjg reviewed (commented):
LGTM, but I just did a quick skim. @erin-le can you review closely?

One thing I'll say is that eventually we need to specialize these (and other) exceptions into some gem5-specific exceptions. (E.g., `GEM5JSONFormatException`)

Harshil2107 committed:
resources: Update error messages

erin-le commented on tests/pyunit/stdlib/resources/pyunit_obtain_resources_check.py:
```diff
@@ -172,3 +172,100 @@ def test_obtain_resources_with_version_invalid_version(
+            in str(context.exception)
+        )
+
+    def test_obtain_resoruces_with_missing_category(self, mock_create_client):
```
```suggestion
    def test_obtain_resources_with_missing_category(self, mock_create_client):
```

erin-le commented on tests/pyunit/stdlib/resources/pyunit_obtain_resources_check.py:
```diff
@@ -172,3 +172,100 @@ def test_obtain_resources_with_version_invalid_version(
+            in str(context.exception)
+        )
+
+    def test_obtain_resources_with_missing_resoruces_in_worklaod(
```
```suggestion
    def test_obtain_resources_with_missing_resources_in_workload(
```

erin-le commented on tests/pyunit/stdlib/resources/pyunit_obtain_resources_check.py:
```diff
@@ -172,3 +172,100 @@ def test_obtain_resources_with_version_invalid_version(
+            in str(context.exception)
+        )
+
+    def test_obtain_resources_with__no_id_in_workloads_in_suite(
```
```suggestion
    def test_obtain_resources_with_no_id_in_workloads_in_suite(
```

erin-le commented on tests/pyunit/stdlib/resources/pyunit_obtain_resources_check.py:
```diff
@@ -172,3 +172,100 @@ def test_obtain_resources_with_version_invalid_version(
+            in str(context.exception)
+        )
+
+    def test_obtain_resources_with__no_input_group_in_workloads_in_suite(
```
```suggestion
    def test_obtain_resources_with_no_input_group_in_workloads_in_suite(
```

erin-le commented on src/python/gem5/resources/downloader.py:
```diff
@@ -261,6 +261,11 @@ def get_resource(
+            if "md5sum" not in resource_json:
+                raise Exception(
+                    f"The resource '{resource_json['id']}' and "
+                    f"'{resource_json['resoruce_version']}' does not contain an md5sum"
```
```suggestion
                    f"'{resource_json['resource_version']}' does not contain an md5sum"
```

erin-le commented on src/python/gem5/resources/resource.py:
```diff
@@ -1169,6 +1218,7 @@ def _get_workload(
             "additional_params" in workload.keys()
             and workload["additional_params"]
         ):
+            assert isinstance(workload["additional_params"], dict)
```
I tested this and got the following for an error message. You may want to add a more descriptive message.

```txt
AssertionError: <EMPTY MESSAGE>
```

erin-le replied:
I used the following to test this:

```json
    {
        "category": "workload",
        "id": "test-invalid-additional-params",
        "description": "A full boot of Ubuntu 24.04 with Linux 6.6.33 for RISCV. It runs an `m5 exit` command when the boot is completed unless the readfile is specified. If specified the readfile will be executed after booting. This workload throws 3 m5 exit events, the events are thrown when kernel is booted, when `after_boot.sh` script is executed and after readfile is run.",
        "function": "set_kernel_disk_workload",
        "resources": {
            "bootloader": {
                "id": "riscv-bootloader-opensbi-1.3.1",
                "resource_version": "1.0.0"
            },
            "kernel": {
                "id": "riscv-linux-6.6.33-kernel",
                "resource_version": "1.0.0"
            },
            "disk_image": {
                "id": "riscv-ubuntu-24.04-img",
                "resource_version": "1.0.0"
            }
        },
        "additional_params": ["additional", "params"],
        "architecture": "RISCV",
        "resource_version": "1.0.0",
        "gem5_versions": [
            "24.0",
            "24.1"
        ]
    },
```

erin-le replied:
After Bobby's changes (PR has 6 commits total), I got the following error message:

```txt
warn: Resource riscv-bootloader-opensbi-1.3.1 version 1.0.0 contains unexpected parameter '_id' that is not used by bootloader. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-bootloader-opensbi-1.3.1 version 1.0.0 contains unexpected parameter 'category' that is not used by bootloader. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-bootloader-opensbi-1.3.1 version 1.0.0 contains unexpected parameter 'author' that is not used by bootloader. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-bootloader-opensbi-1.3.1 version 1.0.0 contains unexpected parameter 'license' that is not used by bootloader. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-bootloader-opensbi-1.3.1 version 1.0.0 contains unexpected parameter 'source_url' that is not used by bootloader. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-bootloader-opensbi-1.3.1 version 1.0.0 contains unexpected parameter 'tags' that is not used by bootloader. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-bootloader-opensbi-1.3.1 version 1.0.0 contains unexpected parameter 'example_usage' that is not used by bootloader. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-bootloader-opensbi-1.3.1 version 1.0.0 contains unexpected parameter 'gem5_versions' that is not used by bootloader. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-bootloader-opensbi-1.3.1 version 1.0.0 contains unexpected parameter 'url' that is not used by bootloader. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-bootloader-opensbi-1.3.1 version 1.0.0 contains unexpected parameter 'is_tar_archive' that is not used by bootloader. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-bootloader-opensbi-1.3.1 version 1.0.0 contains unexpected parameter 'is_zipped' that is not used by bootloader. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-bootloader-opensbi-1.3.1 version 1.0.0 contains unexpected parameter 'size' that is not used by bootloader. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-bootloader-opensbi-1.3.1 version 1.0.0 contains unexpected parameter 'md5sum' that is not used by bootloader. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-linux-6.6.33-kernel version 1.0.0 contains unexpected parameter '_id' that is not used by kernel. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-linux-6.6.33-kernel version 1.0.0 contains unexpected parameter 'category' that is not used by kernel. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-linux-6.6.33-kernel version 1.0.0 contains unexpected parameter 'author' that is not used by kernel. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-linux-6.6.33-kernel version 1.0.0 contains unexpected parameter 'license' that is not used by kernel. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-linux-6.6.33-kernel version 1.0.0 contains unexpected parameter 'source_url' that is not used by kernel. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-linux-6.6.33-kernel version 1.0.0 contains unexpected parameter 'tags' that is not used by kernel. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-linux-6.6.33-kernel version 1.0.0 contains unexpected parameter 'example_usage' that is not used by kernel. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-linux-6.6.33-kernel version 1.0.0 contains unexpected parameter 'gem5_versions' that is not used by kernel. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-linux-6.6.33-kernel version 1.0.0 contains unexpected parameter 'url' that is not used by kernel. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-linux-6.6.33-kernel version 1.0.0 contains unexpected parameter 'is_zipped' that is not used by kernel. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-linux-6.6.33-kernel version 1.0.0 contains unexpected parameter 'is_tar_archive' that is not used by kernel. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-linux-6.6.33-kernel version 1.0.0 contains unexpected parameter 'size' that is not used by kernel. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-linux-6.6.33-kernel version 1.0.0 contains unexpected parameter 'md5sum' that is not used by kernel. Valid parameters are: architecture, description, id, local_path, resource_version, source
warn: Resource riscv-ubuntu-24.04-img version 1.0.0 contains unexpected parameter '_id' that is not used by disk-image. Valid parameters are: description, id, local_path, resource_version, root_partition, source
warn: Resource riscv-ubuntu-24.04-img version 1.0.0 contains unexpected parameter 'category' that is not used by disk-image. Valid parameters are: description, id, local_path, resource_version, root_partition, source
warn: Resource riscv-ubuntu-24.04-img version 1.0.0 contains unexpected parameter 'architecture' that is not used by disk-image. Valid parameters are: description, id, local_path, resource_version, root_partition, source
warn: Resource riscv-ubuntu-24.04-img version 1.0.0 contains unexpected parameter 'size' that is not used by disk-image. Valid parameters are: description, id, local_path, resource_version, root_partition, source
warn: Resource riscv-ubuntu-24.04-img version 1.0.0 contains unexpected parameter 'tags' that is not used by disk-image. Valid parameters are: description, id, local_path, resource_version, root_partition, source
warn: Resource riscv-ubuntu-24.04-img version 1.0.0 contains unexpected parameter 'is_zipped' that is not used by disk-image. Valid parameters are: description, id, local_path, resource_version, root_partition, source
warn: Resource riscv-ubuntu-24.04-img version 1.0.0 contains unexpected parameter 'md5sum' that is not used by disk-image. Valid parameters are: description, id, local_path, resource_version, root_partition, source
warn: Resource riscv-ubuntu-24.04-img version 1.0.0 contains unexpected parameter 'url' that is not used by disk-image. Valid parameters are: description, id, local_path, resource_version, root_partition, source
warn: Resource riscv-ubuntu-24.04-img version 1.0.0 contains unexpected parameter 'code_examples' that is not used by disk-image. Valid parameters are: description, id, local_path, resource_version, root_partition, source
warn: Resource riscv-ubuntu-24.04-img version 1.0.0 contains unexpected parameter 'license' that is not used by disk-image. Valid parameters are: description, id, local_path, resource_version, root_partition, source
warn: Resource riscv-ubuntu-24.04-img version 1.0.0 contains unexpected parameter 'author' that is not used by disk-image. Valid parameters are: description, id, local_path, resource_version, root_partition, source
warn: Resource riscv-ubuntu-24.04-img version 1.0.0 contains unexpected parameter 'source_url' that is not used by disk-image. Valid parameters are: description, id, local_path, resource_version, root_partition, source
warn: Resource riscv-ubuntu-24.04-img version 1.0.0 contains unexpected parameter 'gem5_versions' that is not used by disk-image. Valid parameters are: description, id, local_path, resource_version, root_partition, source
warn: Resource riscv-ubuntu-24.04-img version 1.0.0 contains unexpected parameter 'example_usage' that is not used by disk-image. Valid parameters are: description, id, local_path, resource_version, root_partition, source
AssertionError: <EMPTY MESSAGE>

```

erin-le replied:
I pushed a commit that adds a message to the AssertionError.

erin-le commented on src/python/gem5/resources/resource.py:
```diff
@@ -1191,6 +1241,17 @@ def _get_to_path_and_downloader_partial(
     quiet: bool,
 ) -> Tuple[str, Optional[partial]]:
+
+    if "resource_version" not in resource_json:
```
I tried removing resource version and got the following error:
```
Exception thrown while getting resources '[<gem5.resources.client_api.client_query.ClientQuery object at 0x70d204d17f20>]' from client 'GEM5_RESOURCE_JSON_APPEND'

KeyError: 'resource_version'

At:
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/client_api/jsonclient.py(101): filter_resource
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/client_api/jsonclient.py(121): <lambda>
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/client_api/jsonclient.py(126): get_resources
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/client_api/abstract_client.py(166): get_resources_by_id
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/client.py(336): _get_all_resources_by_id
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/client.py(275): _get_resource_json_obj_from_client
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/client.py(183): get_resource_json_obj
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/resource.py(958): obtain_resource
  arm-hello.py(104): <module>
  /home/bees/gem5-2nd-worktree/src/python/m5/main.py(675): main

```
It looks like the above error occurs before this exception.

erin-le replied:
It appears that if you have the following in your JSON (a resource lacking a resource version field), you will get the same error even if you try to use a different resource.
```json
{
    "category": "binary",
    "id": "test-missing-resource-version",
    "description": "test description for missing category test",
    "architecture": "ARM",
    "is_zipped": false,
    "md5sum": "71b2cb004fe2cda4556f0b1a38638af6",
    "url": "http://dist.gem5.org/dist/develop/test-progs/hello/bin/arm/linux/hello64-static",
    "source": "src/test-source",
    "gem5_versions": [
        "develop",
        "24.0",
        "24.1"
    ]
},
```

erin-le replied:
After Bobby's changes, I got the following error message:

```
Exception thrown while getting resources '[<gem5.resources.client_api.client_query.ClientQuery object at 0x7285c44888f0>]' from client 'GEM5_RESOURCE_JSON_APPEND'

TypeError: string indices must be integers, not 'str'

At:
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/client_api/jsonclient.py(128): filter_resource
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/client_api/jsonclient.py(148): <lambda>
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/client_api/jsonclient.py(153): get_resources
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/client_api/abstract_client.py(170): get_resources_by_id
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/client.py(336): _get_all_resources_by_id
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/client.py(275): _get_resource_json_obj_from_client
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/client.py(183): get_resource_json_obj
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/resource.py(959): obtain_resource
  arm-hello.py(125): <module>
  /home/bees/gem5-2nd-worktree/src/python/m5/main.py(675): main

```

Harshil2107 replied:
I pushed a change that should add checks for resource_version.

erin-le commented on src/python/gem5/resources/resource.py:
```diff
@@ -1191,6 +1241,17 @@ def _get_to_path_and_downloader_partial(
+        raise Exception(
+            f"Resource {resource_json['id']} does not contain a resource version"
+        )
+    if not isinstance(resource_json["resource_version"], str):
```
I tried using the following 

```json
    {
        "category": "binary",
        "id": "test-invalid-resource-version",
        "description": "test description for missing category test",
        "architecture": "ARM",
        "is_zipped": false,
        "md5sum": "71b2cb004fe2cda4556f0b1a38638af6",
        "url": "http://dist.gem5.org/dist/develop/test-progs/hello/bin/arm/linux/hello64-static",
        "source": "src/test-source",
        "resource_version": 1,
        "gem5_versions": [
            "develop",
            "24.0",
            "24.1"
        ]
    },
```

and got this error:

```
Exception: Resource with ID 'test-invalid-resource-version' not found.

At:
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/client.py(280): _get_resource_json_obj_from_client
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/client.py(183): get_resource_json_obj
  /home/bees/gem5-2nd-worktree/src/python/gem5/resources/resource.py(958): obtain_resource
  arm-hello.py(104): <module>
  /home/bees/gem5-2nd-worktree/src/python/m5/main.py(675): main
```
I double checked the error by changing `"resource_version": 1,` to `"resource_version": "1.0.0",` without changing anything else, and observed that the simulation ran successfully after the change.

erin-le replied:
After Bobby's changes, I observed the same thing as above. When running with a valid resource version, 5 additional warnings about unexpected parameters that aren't used by binary are printed.

Harshil2107 replied:
I pushed a change that should add checks for resource_version.

erin-le commented on tests/pyunit/stdlib/resources/refs/obtain-resource.json:
```diff
@@ -63,5 +63,112 @@
+    {
+        "category": "workload",
+        "id": "test-workload-no-resources-field",
+        "description": "test description of missng field in workload",
```
```suggestion
        "description": "test description of missing field in workload",
```

Harshil2107 committed:
misc: Fix spelling errors

BobbyRBruce committed:
resources,tests: Refactor resource schema validator

1. Added the `_resource_schema_validator` function to "resources.py".
   This validates the schema of a gem5 resource JSON/Dict using the
   corresponding Resource object as determined by the value of the
   JSON/Dict's `category` field.
2. Simplified `_get_suite` and `_get_workload`. These functions now
   refector the resource JSON/dict to correspond with `SuiteResource`
   and `WorkloadResource` objects.
3. Updated the behavior of md5sum. If not specified in the resources
   JSON/Dict gem5 will always download the resource and not cache.
4. Rewrote some Exception messages to be clearer.
5. Updated the tests for Resources to use `self.assertEquals` where
   appropriate (`self.assertTrue` was being used).

BobbyRBruce commented:
@Harshil2107 I've made some changes to the way we check the schema and some minor~ish refactoring. See the last commit in the chain and let me know what you think.

Also, did you address @erin-le 's comments?

erin-le committed:
python: fix CI test failures for JSON schema issues exceptions

This commit fixes test failures for PR 1912, which adds exceptions
and error checking for schema issues in resource JSON files. The
additional_params key previously would not be removed from the
workload if it had no value, which would incorrectly trigger an
assert.

Harshil2107 commented:
I will take a look at the changes, an address erin's requested changes.

erin-le committed:
python: Add assert message for invalid additional params

This commit adds an assert message for when additional params
are not a dict.

Harshil2107 committed:
resources: remove check for non required parameters

- The JSON object for resources contain fields that may not be required by gem5 but they might be required by the resources website or provide more information about the resource. This change removes the warnings in the validator that are shown if the resource contains extra fields.

Harshil2107 committed:
resources: Add checks for resource_version when getting resources

Harshil2107 committed:
resources: Add a check to make sure the resources.json is a list

erin-le committed:
resources: make resource download reason more explicit

Addresses Issue 1934. This commit adds additional messages
that are printed when a resource is re-downloaded. More
specifically, new messages have been added for when the version
of the cached local resource doesn't match with the version of
the requested resource, and for when the md5sum doesn't match.

erin-le reviewed (approved):
Testing approval

powerjg commented:
Should we merge commit or squash?

Harshil2107 commented:
I think We should squash merge as initial 5 commits of this PR are overridden by Bobby's changes. I can try to rebase the commits if we want to merge instead.
